RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc

```diff
@@ -98,9 +98,151 @@ namespace {
     const tensorflow::Session* session_;
   };
 
+  class TfDnnBatch {
+  public:
+    TfDnnBatch(const edm::ParameterSet& cfg, edm::ConsumesCollector iC)
+        : tfDnnLabel_(cfg.getParameter<std::string>("tfDnnLabel")),
+          tfDnnToken_(iC.esConsumes(edm::ESInputTag("", tfDnnLabel_))),
+          session_(nullptr),
+          bsize_(cfg.getParameter<int>("batchSize"))
+
+    {}
+
+    static const char* name() { return "trackTfClassifierDefaultBatch"; }
+
+    static void fillDescriptions(edm::ParameterSetDescription& desc) {
+      desc.add<std::string>("tfDnnLabel", "trackSelectionTf");
+      desc.add<int>("batchSize", 16);
+    }
+    void beginStream() {}
+
+    void initEvent(const edm::EventSetup& es) {
+      if (session_ == nullptr) {
+        session_ = es.getData(tfDnnToken_).getSession();
+      }
+    }
+
+    std::vector<float> operator()(reco::TrackCollection const& tracks,
+                                  reco::BeamSpot const& beamSpot,
+                                  reco::VertexCollection const& vertices) const {
+      //std::cout <<"in batch eval part"<< std::endl;
+
+      int size_in = (int)tracks.size();
+      int bsize = bsize_;
+      int nbatches = size_in / bsize;
+
+      std::vector<float> output;
+      output.resize(size_in);
+
+      tensorflow::Tensor input1(tensorflow::DT_FLOAT, {bsize, 29});
+      tensorflow::Tensor input2(tensorflow::DT_FLOAT, {bsize, 1});
+
+      for (auto nb = 0; nb < nbatches + 1; nb++) {
+        for (auto nt = 0; nt < bsize; nt++) {
+          int itrack = nt + bsize * nb;
+          if (itrack >= size_in)
+            continue;
+          const auto& trk = tracks.at(itrack);
+
+          const auto& bestVertex = getBestVertex(trk, vertices);
+
+          input1.matrix<float>()(nt, 0) = trk.pt();
+          input1.matrix<float>()(nt, 1) = trk.innerMomentum().x();
+          input1.matrix<float>()(nt, 2) = trk.innerMomentum().y();
+          input1.matrix<float>()(nt, 3) = trk.innerMomentum().z();
+          input1.matrix<float>()(nt, 4) = trk.innerMomentum().rho();
+          input1.matrix<float>()(nt, 5) = trk.outerMomentum().x();
+          input1.matrix<float>()(nt, 6) = trk.outerMomentum().y();
+          input1.matrix<float>()(nt, 7) = trk.outerMomentum().z();
+          input1.matrix<float>()(nt, 8) = trk.outerMomentum().rho();
+          input1.matrix<float>()(nt, 9) = trk.ptError();
+          input1.matrix<float>()(nt, 10) = trk.dxy(bestVertex);
+          input1.matrix<float>()(nt, 11) = trk.dz(bestVertex);
+          input1.matrix<float>()(nt, 12) = trk.dxy(beamSpot.position());
+          input1.matrix<float>()(nt, 13) = trk.dz(beamSpot.position());
```

Comment:
can this be wrapped in a function call? The arguments for the function would be `(input1.matrix<float>(), nt, trk)`.
This will be called in the non-batch variant and would avoid code replication.

In the end, I think it's even more practical to replace `TfDnn` and avoid additional copy-pastes in the remaining boilerplate code